src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,185 @@
+            Node<T> newNode = new Node<>(currentNode.prev, value, currentNode);
+            currentNode.prev.next = newNode;
+            currentNode.prev = newNode;
+            size++;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T item: list) {
+            add(item);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        return getNodeByIndex(index).item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        Node<T> node = getNodeByIndex(index);
+        T oldValue = node.item;
+        node.item = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        Node<T> node = getNodeByIndex(index);
+        unlink(node);
+        size--;
+        return node.item;
     }
 
     @Override
     public boolean remove(T object) {
-        return false;
+        Node<T> node = head;
+        while (node != null) {
+            if (node.item == null && object == null) {
+                break;
+            }
+            if (node.item != null && node.item.equals(object)) {
+                break;
+            }
+            node = node.next;
+        }
+        if (node == null) {
+            return false;
+        }
+        unlink(node);
+        size--;
+        return true;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private static class Node<T> {
+        private T item;
+        private Node<T> next;
+        private Node<T> prev;
+
+        Node(Node<T> prev, T item, Node<T> next) {
+            this.item = item;
+            this.prev = prev;
+            this.next = next;
+        }
+    }
+
+    private void checkIndex(int index) {
+        if (index >= size || index < 0) {
+            throw new IndexOutOfBoundsException("Specified index is invalid");
+        }
+    }
+
+    private Node<T> searchFromStart(int index) {
+        Node<T> item = head;
+        int counter = 0;
+        do {
+            if (counter == index) {
+                return item;
+            }
+            item = item.next;
+            counter++;
+        } while (counter <= index);
+        return item;
+    }
+
+    private Node<T> searchFromEnd(int index) {
+        Node<T> item = tail;
+        int counter = size - 1;
+        do {
+            if (counter == index) {
+                return item;
+            }
+            item = item.prev;
+            counter--;
+        } while (counter >= index);
+        return item;
+    }
+
+    private Node<T> getNodeByIndex(int index) {
+        checkIndex(index);
+        int listMiddleIndex = size / 2;
+        if (index <= listMiddleIndex) {
+            return searchFromStart(index);
+        } else {
+            return searchFromEnd(index);
+        }
+    }
```

Comment:
we should place methods in order of their calling to make reading more simple (from up to down)
also, put these methods right after last `remove()` method, as this is the last place where `getNodeByIndex` being called
```suggestion
    private Node<T> getNodeByIndex(int index) {
        checkIndex(index);
        int listMiddleIndex = size / 2;
        if (index <= listMiddleIndex) {
            return searchFromStart(index);
        } else {
            return searchFromEnd(index);
        }
    }
    
    private Node<T> searchFromStart(int index) {
        Node<T> item = head;
        int counter = 0;
        do {
            if (counter == index) {
                return item;
            }
            item = item.next;
            counter++;
        } while (counter <= index);
        return item;
    }

    private Node<T> searchFromEnd(int index) {
        Node<T> item = tail;
        int counter = size - 1;
        do {
            if (counter == index) {
                return item;
            }
            item = item.prev;
            counter--;
        } while (counter >= index);
        return item;
    }
```